p2p: peer connection bug fixes

jonatack opened the pull request:
This pull fixes several peer connection bugs in our p2p code, along with the logging that uncovered them:

1. Fix detection of inbound peer connections in `GetAddedNodeInfo`.

2. Fix addnode CJDNS peers not detected in `GetAddedNodeInfo`, causing `ThreadOpenAddedConnections` to continually retry to connect to them and RPC `getaddednodeinfo` incorrectly showing them as not connected.

3. Fix `ThreadOpenConnections` not detecting inbound CJDNS connections and making automatic outbound connections to them.

4. Fix detection of already connected peers in `AlreadyConnectedToAddress()`.

5. Fix detection of already connected peers when making outbound connections in `ConnectNode`.

6. Do not accept inbound connections in `CreateNodeFromAcceptedSocket` from I2P peers we're already connected to, as building I2P tunnels is expensive.

7. Fix making automatic outbound connections in `ThreadOpenConnections` to addnode peers, in order not to allocate our limited outbound slots to them and to ensure addnode connections benefit from their intended protections. Our addnode logic usually connects the addnode peers before the automatic outbound logic does, but not always, as a connection race can occur (see the commit message for further details and mainnet examples). When an addnode peer is connected as an automatic outbound peer and is the only connection we have to a network, it can be protected by our new outbound network-specific eviction logic and persist in the "wrong role". Fix these issues by checking if the selected address is an addnode peer in our automatic outbound connection logic.

Update the p2p logging with the improvements that allowed seeing/understanding/debugging the current behavior. Please see the commit messages for details.

Simplify `MaybePickPreferredNetwork` to return `std::optional`, make it a const class method, and add Clang thread-safety analysis annotation and related assertions.

jonatack committed:
refactor: simplify MaybePickPreferredNetwork, drop out-param, return optional

and make it a const class method, and add Clang thread-safety analysis annotation and
related assertions.

jonatack committed:
p2p: log when protecting last remaining network peer from eviction

This logging has been helpful for observing and debugging our peer connection
and eviction behavior.

Also improve related log entries for
- eviction of an inbound peer
- protecting a peer from eviction for a bad or slow chain

dergoegge commented:
The title "p2p: outbound network diversity improvements" makes it sound like you are improving on the logic for making outbound connections but really you are refactoring and adding additional logging. Would you mind making the title more accurate?

> Please see the commit messages for details.

The PR description is added to the merge commit, so in my opinion it makes sense to have at least a summary of what the PR does in the description.

From our [docs](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#creating-the-pull-request):
```
The body of the pull request should contain sufficient description of what the patch does, and even more importantly, why, with 
justification and reasoning. You should include references to any discussions (for example, other issues or mailing list
discussions).
```

ajtowns commented:
This mostly seems like pointless refactoring (changing a `switch` to an `||` of two functions that themselves have switches). Doesn't seem worth the review effort to me.

jonatack committed:
p2p, bugfix: correctly detect already connected peers in ConnectNode()

by upgrading the checks from FindNode() to AlreadyConnectedToAddress().

jonatack committed:
p2p: improve inbound/outbound/anchor/network-specific log messages

jonatack committed:
p2p, bugfix: do not make automatic outbound connections to addnode peers

to allocate our limited outbound slots correctly, and to ensure addnode
connections benefit from their intended protections.

Our addnode logic usually connects the addnode peers before the automatic
outbound logic does, but not always, as a connection race can occur.

For instance, our internet connection or router (or the addnode peer) could be
temporarily offline, and then return online during the automatic outbound
thread.  Or we could add a new manual peer using the `addnode` RPC at that time.
The race can be more apparent when our node doesn't know many peers, or with
networks like cjdns that currently have few bitcoin peers.

Examples on mainnet using logging added in the same pull request:

2023-08-12T14:51:05.681743Z [opencon] [net.cpp:1949] [ThreadOpenConnections]
[net:debug] Not making automatic network-specific outbound-full-relay connection
to i2p peer selected for manual (addnode) connection: [geh...odq.b32.i2p]:0

2023-08-13T03:59:28.050853Z [opencon] [net.cpp:1949] [ThreadOpenConnections]
[net:debug] Not making automatic block-relay-only connection to onion peer
selected for manual (addnode) connection: kpg...aid.onion:8333

2023-08-13T16:21:26.979052Z [opencon] [net.cpp:1949] [ThreadOpenConnections]
[net:debug] Not making automatic network-specific outbound-full-relay connection
to cjdns peer selected for manual (addnode) connection: [fcc...8ce]:8333

2023-08-14T20:43:53.401271Z [opencon] [net.cpp:1949] [ThreadOpenConnections]
[net:debug] Not making automatic network-specific outbound-full-relay connection
to cjdns peer selected for manual (addnode) connection: [fc7...59e]:8333

2023-08-15T00:10:01.894147Z [opencon] [net.cpp:1949] [ThreadOpenConnections]
[net:debug] Not making automatic feeler connection to i2p peer selected for
manual (addnode) connection: geh...odq.b32.i2p:8333

When an addnode peer is connected as an automatic outbound peer and is the only
connection we have to a network, it can be protected by our new outbound
eviction logic and persist in the "wrong role".

Finally, there does not seem to be a reason to make block-relay or short-lived
feeler connections to addnode peers, as the addnode logic will ensure we connect
to them if they are up, within the addnode connection limit.

Fix these issues by checking if the address is an addnode peer in our automatic
outbound connection logic.

Co-authored-by: Vasil Dimov <vd@FreeBSD.org>

jonatack committed:
p2p, bugfix: detect addnode cjdns peers in GetAddedNodeInfo()

Addnode (manual) peers connected to us via the cjdns network are currently not
detected by CConnman::GetAddedNodeInfo(), i.e. fConnected is always false.

This causes the following issues:

- RPC `getaddednodeinfo` incorrectly shows them as not connected

- CConnman::ThreadOpenAddedConnections() continually retries to connect them

jonatack committed:
p2p, bugfix: detect inbound cjdns peers in AlreadyConnectedToAddress()

Currently ThreadOpenConnections() does not detect inbound cjdns peers that we
are already connected to when it calls AlreadyConnectedToAddress().

Fix this by calling MaybeFlipIPv6toCJDNS before the search.

vasild commented on src/net.cpp:
```diff
@@ -1928,6 +1921,25 @@ void CConnman::ThreadOpenConnections(const std::vector<std::string> connect)
                 continue;
             }
 
+            // Do not make automatic outbound connections to addnode peers, to
+            // avoid using our limited outbound slots for them and to ensure
+            // addnode connections benefit from their intended protections.
+            {
+                const auto str_addr{addr.ToStringAddr()};
+                const auto str_addr_port{addr.ToStringAddrPort()};
+
+                LOCK(m_added_nodes_mutex);
+                if (m_added_nodes.size() < 16 // bound the query to a reasonable limit
+                    && std::any_of(m_added_nodes.cbegin(), m_added_nodes.cend(), [&](const auto& a) { return a == str_addr || a == str_addr_port; })) {
```
Good catch with this deficiency!

If `m_added_nodes` is changed to `std::unordered_set`, then this can be a simple `m_added_nodes.count() > 0` and `CConnman::AddNode()` and `CConnman::RemoveAddedNode()` could be simplified (and made faster since they are now `O(number of added nodes)` and will become `O(1)`, I guess performance is irrelevant in practice because of the small number of elements).

Maybe this is worth a separate PR since it is distinct from other commits (one commit to change it to `std::unordered_set` and one to add this check).

jonatack committed:
refactor: change m_added_node_params from vector to unordered_set

which simplifies and speeds up the AddNode() and RemoveAddedNode() methods,
as well as the next commit, "p2p: do not make automatic outbound connections
to addnode peers", which checks if an address is already in m_added_node_params.

This refactoring is covered by functional tests in rpc_net.py.

Co-authored-by: Vasil Dimov <vd@FreeBSD.org>

vasild commented on src/net.cpp:
```diff
@@ -408,7 +397,8 @@ CNode* CConnman::FindNode(const CService& addr)
 
 bool CConnman::AlreadyConnectedToAddress(const CAddress& addr)
 {
-    return FindNode(static_cast<CNetAddr>(addr)) || FindNode(addr.ToStringAddrPort());
+    const CService service{MaybeFlipIPv6toCJDNS(addr)};
+    return FindNode(service) || FindNode(service.ToStringAddrPort());
```
Before this it would compare just the address and after this change it would compare the address and the port. I think that is undesirable because for inbound connections we see the peer as e.g. `1.2.3.4:somerandomport` and when we try to connect to `1.2.3.4:8333` we should assume we are already connected to that peer.

vasild commented on src/net.cpp:
```diff
@@ -1995,7 +2004,7 @@ std::vector<AddedNodeInfo> CConnman::GetAddedNodeInfo() const
     }
 
     for (const std::string& strAddNode : lAddresses) {
-        CService service(LookupNumeric(strAddNode, Params().GetDefaultPort(strAddNode)));
+        CService service{MaybeFlipIPv6toCJDNS(LookupNumeric(strAddNode, Params().GetDefaultPort(strAddNode)))};
```
The existent code in `master` seems to be comparing the ports. Is there another bug that an inbound connection from `1.2.3.4` would not be considered as "we are connected to `1.2.3.4:8333`"?

vasild replied:
I mean, if we already have an inbound connection from `1.2.3.4` (ie we are connected to `1.2.3.4:somerandomport`) and this code checks whether we are connected to `1.2.3.4:8333` it would wrongly assume that we are not and would open an outbound connection to that address.

vasild commented on src/net.cpp:
```diff
@@ -2621,24 +2630,15 @@ std::vector<CAddress> CConnman::GetAddresses(CNode& requestor, size_t max_addres
-    }
-
-    m_added_nodes.push_back(strNode);
+    if (m_added_nodes.count(strNode)) return false;
+    m_added_nodes.insert(strNode);
     return true;
```
Or this:

```cpp
return m_added_nodes.insert(strNode).second;
```

jonatack committed:
p2p: improve ConnectNode log, rm redundant ThreadOpenConnections log

By moving the "Trying connection" log entry in ConnectNode to after the address
resolution, we can:

- not needlessly log the attempt until after checking if we are already connected to it

- know the network the peer connected through, i.e. for cjdns peers

- add useful peer info to the log entry for debugging purposes

This change makes the feeler log entry in CConnman::ThreadOpenConnections()
redundant...

2023-09-06T01:11:42.614464Z Making feeler connection to dft...dyd.onion:8333
2023-09-06T01:11:42.614722Z Trying feeler connection to net=onion, peeraddr=dft...dyd.onion:8333, lastseen=4.2hrs

...so remove it.

vasild committed:
p2p, bugfix: detect all inbound connections in GetAddedNodeInfo()

Due to their potentially different port numbers, inbound peers can be overlooked
in CConnman::GetAddedNodeInfo() when comparing CServices, i.e. both addr and port.

Fix this by comparing CNetAddr instances, i.e. addr only, for inbound peers.

Co-authored-by: Jon Atack <jon@atack.com>

jonatack committed:
p2p: do not accept inbound connections from already connected I2P peers

as creating I2P tunnels is expensive.

These duplicate connection requests come most frequently from inbound peers.
The other fixes in this PR may mitigate this behavior by our peers, but not for
nodes running earlier versions of Bitcoin Core.

Extend a related functional test.

mzumsande commented on src/net.cpp:
```diff
@@ -1814,6 +1814,16 @@ void CConnman::CreateNodeFromAcceptedSocket(std::unique_ptr<Sock>&& sock,
+    // Don't accept inbound connections from peers already connected to us.
+    // Skip this check for Tor, as their inbound/outbound addresses differ.
+    if (!inbound_onion) {
+        if (auto pnode = AlreadyConnectedToAddress(addr)) {
```
Similar question here: Could this make some local network testing setups no longer viable?

mzumsande reviewed (commented):
Just had a short look and some questions, will review properly in the next weeks.

dergoegge commented:
If you're fixing bugs, it would be very beneficial to also add tests at the same time.

jonatack committed:
refactor: pass AddedNodeParams to CConnman::RemoveAddedNode()

instead of a std::string (and rename its argument and that of
CConnman::AddNode() to "params").

Passing AddedNodeParams makes the two member functions AddNode and
RemoveAddedNode consistent, and is needed in order to refactor
`m_added_node_params` from a vector to an unordered set in the next commit.

vasild commented on src/net.cpp:
```diff
@@ -2734,13 +2734,13 @@ std::vector<AddedNodeInfo> CConnman::GetAddedNodeInfo() const
 
     // Build a map of all already connected addresses (by IP:port and by name) to inbound/outbound and resolved CService
-    std::map<CService, bool> mapConnected;
+    std::map<CNetAddr, bool> mapConnected;
```
Right. Comparing ports for incoming connections is problematic also. I think the right approach for determining if we are connected to `addr:port`:

* For incoming connections, compare just `addr` and ignore `port`
* For outgoing connections, compare both `addr` and `port`.

This can't be done with a single map of either `CService` or `CNetAddr`. I guess two containers are needed, as this:

<details>
<summary>[patch] compare differently inbound and outbound addresses</summary>

```diff
diff --git i/src/net.cpp w/src/net.cpp
index 91282d927d..0eb0023112 100644
--- i/src/net.cpp
+++ w/src/net.cpp
@@ -2809,41 +2809,49 @@ std::vector<AddedNodeInfo> CConnman::GetAddedNodeInfo() const
         LOCK(m_added_nodes_mutex);
         ret.reserve(m_added_node_params.size());
         std::copy(m_added_node_params.cbegin(), m_added_node_params.cend(), std::back_inserter(lAddresses));
     }
 
 
-    // Build a map of all already connected addresses (by IP:port and by name) to inbound/outbound and resolved CService
-    std::map<CNetAddr, bool> mapConnected;
+    // Collect all addresses to which we are already connected.
+    std::unordered_set<CNetAddr, CNetAddrHash> connected_to_inbound;
+    std::unordered_set<CService, CServiceHash> connected_to_outbound;
     std::map<std::string, std::pair<bool, CService>> mapConnectedByName;
     {
         LOCK(m_nodes_mutex);
         for (const CNode* pnode : m_nodes) {
             if (pnode->addr.IsValid()) {
-                mapConnected[static_cast<CNetAddr>(pnode->addr)] = pnode->IsInboundConn();
+                if (pnode->IsInboundConn()) {
+                    connected_to_inbound.insert(pnode->addr);
+                } else {
+                    connected_to_outbound.insert(pnode->addr);
+                }
             }
             std::string addrName{pnode->m_addr_name};
             if (!addrName.empty()) {
                 mapConnectedByName[std::move(addrName)] = std::make_pair(pnode->IsInboundConn(), static_cast<const CService&>(pnode->addr));
             }
         }
     }
 
     for (const auto& addr : lAddresses) {
         CService service(LookupNumeric(addr.m_added_node, GetDefaultPort(addr.m_added_node)));
         AddedNodeInfo addedNode{addr, CService(), false, false};
         if (service.IsValid()) {
-            // strAddNode is an IP:port
-            auto it = mapConnected.find(static_cast<CNetAddr>(service));
-            if (it != mapConnected.end()) {
+            // addr.m_added_node is an addr:port
+            if (connected_to_outbound.count(service) > 0) {
+                addedNode.resolvedAddress = service;
+                addedNode.fConnected = true;
+                addedNode.fInbound = false;
+            } else if (connected_to_inbound.count(service) > 0) {
                 addedNode.resolvedAddress = service;
                 addedNode.fConnected = true;
-                addedNode.fInbound = it->second;
+                addedNode.fInbound = true;
             }
         } else {
-            // strAddNode is a name
+            // addr.m_added_node is a name
             auto it = mapConnectedByName.find(addr.m_added_node);
             if (it != mapConnectedByName.end()) {
                 addedNode.resolvedAddress = it->second.second;
                 addedNode.fConnected = true;
                 addedNode.fInbound = it->second.first;
             }
diff --git i/src/netaddress.h w/src/netaddress.h
index ad09f16799..99a3b0fc22 100644
--- i/src/netaddress.h
+++ w/src/netaddress.h
@@ -254,12 +254,13 @@ public:
             UnserializeV2Stream(s);
         } else {
             UnserializeV1Stream(s);
         }
     }
 
+    friend class CNetAddrHash;
     friend class CSubNet;
 
 private:
     /**
      * Parse a Tor address and set this object to it.
      * @param[in] addr Address to parse, must be a valid C string, for example
@@ -473,12 +474,36 @@ private:
         // will not be gossiped, but continue reading next addresses from the stream.
         m_net = NET_IPV6;
         m_addr.assign(ADDR_IPV6_SIZE, 0x0);
     }
 };
 
+class CNetAddrHash
+{
+public:
+    CNetAddrHash()
+        : m_salt_k0{GetRand<uint64_t>()},
+          m_salt_k1{GetRand<uint64_t>()}
+    {
+    }
+
+    CNetAddrHash(uint64_t salt_k0, uint64_t salt_k1) : m_salt_k0{salt_k0}, m_salt_k1{salt_k1} {}
+
+    size_t operator()(const CNetAddr& a) const noexcept
+    {
+        CSipHasher hasher(m_salt_k0, m_salt_k1);
+        hasher.Write(a.m_net);
+        hasher.Write(a.m_addr);
+        return static_cast<size_t>(hasher.Finalize());
+    }
+
+private:
+    const uint64_t m_salt_k0;
+    const uint64_t m_salt_k1;
+};
+
 class CSubNet
 {
 protected:
     /// Network (base) address
     CNetAddr network;
     /// Netmask, in network byte order
```
</details>

vasild commented on src/net.cpp:
```diff
@@ -474,15 +483,18 @@ CNode* CConnman::ConnectNode(CAddress addrConnect, const char *pszDest, bool fCo
         }
     }
 
+    LogPrintLevel(BCLog::NET, BCLog::Level::Debug, "Trying v%i %s connection to net=%s%s, lastseen=%.1fhrs\n",
```
nit:
```suggestion
    LogPrintLevel(BCLog::NET, BCLog::Level::Debug, "Trying v%i %s connection to net=%s%s, lastseen=%.1fh ago\n",
```

vasild commented on src/net.cpp:
```diff
@@ -1869,7 +1893,9 @@ void CConnman::CreateNodeFromAcceptedSocket(std::unique_ptr<Sock>&& sock,
     m_msgproc->InitializeNode(*pnode, nodeServices);
 
-    LogPrint(BCLog::NET, "connection from %s accepted\n", addr.ToStringAddrPort());
+    LogPrintLevel(BCLog::NET, BCLog::Level::Debug, "New %s connection accepted from peer=%d, net=%s%s\n",
```
nit: I find "...connection accepted from peer=5"  a bit confusing. Maybe:
```suggestion
    LogPrintLevel(BCLog::NET, BCLog::Level::Debug, "New %s connection accepted, peer=%d, net=%s%s\n",
```

vasild reviewed (commented):
Reviewed as of 2a9db3a16a93d497bfe496b7c35edac975ee81d2. Good fixes! LGTM modulo the comments below.

vasild commented on src/net.h:
```diff
@@ -101,6 +101,17 @@ typedef int64_t NodeId;
+struct AddedNodeHasher {
+    size_t operator()(const AddedNodeParams& p) const noexcept
+    {
+        return std::hash<std::string>()(p.m_added_node) ^ (std::hash<bool>()(p.m_use_v2transport) << 1);
```
I am not sure why they used `<< 1`. It seems to be useless and also it drops one bit from `h2`. Maybe to avoid returning `0` if first name and last name happen to be equal?

I think for the purposes of hashing `m_added_node_params` this suffices:
```cpp
    size_t operator()(const AddedNodeParams& p) const noexcept
    {
        return std::hash<std::string>()(p.m_added_node) ^ p.m_use_v2transport;
    }
```

However `std::hash` is not used anywhere in the code (modulo in some tests and benchmarks). `CServiceHash` uses `CSipHasher`, but there are security concerns with an outsider being able to fiddle with the hashing, maybe this is an overkill for `AddedNodeParams`. If that is to be used here, then it would look like:

```cpp
        CSipHasher hasher(m_salt_k0, m_salt_k1);
        hasher.Write(MakeUCharSpan(p.m_added_node));
        hasher.Write(p.m_use_v2transport);
        return static_cast<size_t>(hasher.Finalize());
```

Or use some of the stateless `Hash*` functions from `src/hash.h`.

I guess @sipa can give us an advice?

vasild commented on src/net.cpp:
```diff
@@ -408,9 +409,32 @@ CNode* CConnman::FindNode(const CService& addr)
+    } else {
+        // For making an outbound connection, check the existing outbound connections
+        // by addr and port, and if none, then inbound connections by addr only.
+        pnode = FindNode(service.ToStringAddrPort(), /*inbound=*/false);
```
Can use `CConnman::FindNode(const CService& addr)` here?
```suggestion
        pnode = FindNode(service, /*inbound=*/false);
```

but then we will not be comparing against `CNode::m_addr_name`. I wonder if we should compare against both `CNode::addr` and `CNode::m_addr_name`?

vasild replied:
Looks good. Can `CNode::addr` ever be invalid?

vasild commented on src/net.cpp:
```diff
@@ -408,9 +409,32 @@ CNode* CConnman::FindNode(const CService& addr)
+        // For making an outbound connection, check the existing outbound connections
+        // by addr and port, and if none, then inbound connections by addr only.
+        pnode = FindNode(service.ToStringAddrPort(), /*inbound=*/false);
+        if (!pnode) pnode = FindNode(static_cast<CNetAddr>(service), /*inbound=*/true);
+    }
+    if (pnode && log) {
+        LogPrintLevel(BCLog::NET, inbound ? BCLog::Level::Debug : BCLog::Level::Info,
+                      "Not %s new %s connection%s, already connected to %s %s %s peer=%d%s, version=%d, subver=%s\n",
+                      inbound ? "accepting" : "opening", ConnectionTypeAsString(conn_type),
+                      fLogIPs ? strprintf(" %s %s", inbound ? "from" : "to", addr.ToStringAddrPort()) : "",
+                      TransportTypeAsString(pnode->m_transport->GetInfo().transport_type),
+                      pnode->ConnectionTypeAsString(), GetNetworkName(pnode->ConnectedThroughNetwork()),
```
Dereferencing `pnode` after `FindNode()` has released `m_nodes_mutex` is not safe. I don't like adding recursive mutex calls, but without altering `FindNode()` we have to do that - this code has to acquire the mutex before any call to `FindNode()`, dereference the returned pointer and then release the mutex:

```diff
 bool CConnman::AlreadyConnectedToAddress(const CAddress& addr, ConnectionType conn_type, bool log)
 {
     if (Params().IsTestChain()) return false; // allow bilateral connections on test chains
     const CService service{MaybeFlipIPv6toCJDNS(addr)};
     const bool inbound{conn_type == ConnectionType::INBOUND};
     CNode* pnode{nullptr};
+    LOCK(m_nodes_mutex);
     if (inbound) {
         // For accepting an inbound connection, check all existing connections by addr only.
         pnode = FindNode(static_cast<CNetAddr>(service));
     } else {
         // For making an outbound connection, check the existing outbound connections
         // by addr and port, and if none, then inbound connections by addr only.
         pnode = FindNode(service.ToStringAddrPort(), /*inbound=*/false);
         if (!pnode) pnode = FindNode(static_cast<CNetAddr>(service), /*inbound=*/true);
     }
     if (pnode && log) {
         LogPrintLevel(BCLog::NET, inbound ? BCLog::Level::Debug : BCLog::Level::Info,
                       "Not %s new %s connection%s, already connected to %s %s %s peer=%d%s, version=%d, subver=%s\n",
                       inbound ? "accepting" : "opening", ConnectionTypeAsString(conn_type),
                       fLogIPs ? strprintf(" %s %s", inbound ? "from" : "to", addr.ToStringAddrPort()) : "",
                       TransportTypeAsString(pnode->m_transport->GetInfo().transport_type),
                       pnode->ConnectionTypeAsString(), GetNetworkName(pnode->ConnectedThroughNetwork()),
                       pnode->GetId(), fLogIPs ? strprintf(", peeraddr=%s", pnode->addr.ToStringAddrPort()) : "",
                       pnode->nVersion.load(), WITH_LOCK(pnode->m_subver_mutex, return pnode->cleanSubVer));
     }
-    return pnode;
+    return pnode != nullptr;
 }
```

vasild reviewed (commented):
Looks good as of 44ceac2ef9 (the last commit needs to be dispersed).

jonatack committed:
refactor: move locks from CConnman::FindNode methods to their callers

Dereferencing the pointer result after FindNode* has released m_nodes_mutex is unsafe,
so we currently have to add unneeded recursive locks before calling FindNode*.

Fix this by acquiring the mutex before calling FindNode* rather than inside them.

Co-authored-by: Vasil Dimov <vd@FreeBSD.org>

jonatack committed:
refactor: allow CConnman::FindNode methods to filter by inbound status

jonatack committed:
p2p, bugfix: correctly detect connected peers in AlreadyConnectedToAddress()

using the following logic:

If we are about to open an outbound connection
  - for all outbound connections, compare the peer's address and port
  - for all inbound connections and provided we're not making a manual connection,
    compare only the address and ignore the port

If we are about to accept an inbound connection
  - for all connections, both inbound and outbound, compare only the address and
    ignore the port

Fix the logging to be actually useful by stating which peer is already connected,
and further improve it to provide full details for debugging purposes.

Co-authored-by: Vasil Dimov <vd@FreeBSD.org>

vasild reviewed (approved):
ACK f196b096f55f909e2bb63d5ff43bc3418c0ebbab

dergoegge commented on src/net.cpp:
```diff
@@ -1819,6 +1819,14 @@ void CConnman::CreateNodeFromAcceptedSocket(std::unique_ptr<Sock>&& sock,
 
     const bool inbound_onion = std::find(m_onion_binds.begin(), m_onion_binds.end(), addr_bind) != m_onion_binds.end();
+
+    // Do not accept inbound connections from peers already connected to us.
```
Just because the connections come from the same IP does no mean it is the same peer, e.g. multiple peers could be connecting to you through the same Tor exit relay.

This does not seem like a bug fix and more like a behavior change. Our inbound eviction logic should handle stuff like this and it probably already does.

dergoegge replied:
> Inbound tor peers are exempted from this check.

We don't have logic for detecting if someone is using Tor as a proxy to connect to the IPv4 network. Our "inbound Tor" classification simply refers to someone connecting to our hidden service.

Another example would be multiple peers connecting to us from the same VPN provider.

vasild replied:
Yes, different peers can be connecting to us through the same NAT box, Tor exit node, or VPN server. Or it might be the same peer indeed. I think that the pros of assuming it is the same peer out-weight the pros of assuming it is not. However, I can see how this can be arguable and without having some solid numbers to back either argument maybe best to keep this unchanged.

fanquake commented on src/net.cpp:
```diff
@@ -328,42 +329,67 @@ bool IsLocal(const CService& addr)
+bool CConnman::AlreadyConnectedToAddress(const CAddress& addr, ConnectionType conn_type, bool log)
 {
-    return FindNode(static_cast<CNetAddr>(addr)) || FindNode(addr.ToStringAddrPort());
+    if (Params().IsTestChain()) return false; // allow bilateral connections on test chains
```
```cpp
    if (Params().IsTestChain()) return false; // allow bilateral connections on test chains
```

This seems very out of place to me, and would also prevent writing certain types of tests?

fanquake replied:
Which tests?

fanquake replied:
> They would need to be updated not to make bilateral connections.

I would suggest updating the tests to account for the behaviour you've changed (it may also help to make the bug fixes clearer for reviewers); rather than introducing per-chain, networking functionality. This seems generally dangerous, and mostly unintuitive.

brunoerg commented:
Concept ACK

fanquake commented on src/net.cpp:
```diff
@@ -2783,18 +2833,21 @@ std::vector<AddedNodeInfo> CConnman::GetAddedNodeInfo() const
 
     for (const auto& addr : lAddresses) {
-        CService service(LookupNumeric(addr.m_added_node, GetDefaultPort(addr.m_added_node)));
+        CService service{MaybeFlipIPv6toCJDNS(LookupNumeric(addr.m_added_node, GetDefaultPort(addr.m_added_node)))};
```
> MaybeFlipIPv6toCJDNS

Is there any way to better encapsulate this `MaybeFlipIPv6toCJDNS()` usage? This seems to be a frequent pattern (a number of additional calls were added in #27071), and it seems like something that would be better handled at a lower level, rather than having to continually remember to wrap various networking calls.

I guess this is also able to remain broken is various ways because we don't have an specific tests for this functionality?

mzumsande commented on src/net.cpp:
```diff
@@ -361,11 +361,38 @@ CNode* CConnman::FindNode(const CService& addr, std::optional<bool> inbound)
 }
 
-bool CConnman::AlreadyConnectedToAddress(const CAddress& addr)
+bool CConnman::AlreadyConnectedToAddress(const CAddress& addr, ConnectionType conn_type, bool log)
```
commit: `p2p, bugfix: correctly detect connected peers in AlreadyConnectedToAddress()`
Could you list some examples of scenarios that are not working correctly currently and are improved by this? I'm trying to understand the impacts of this commit better.

fanquake commented:
> Note also, that item 7 in the PR description involves a regression in v26.x.

What is the regression? Which change caused it?

mzumsande commented:
> Concept ACKs on fixing the bugs would be encouraging 😃

I think that some of the commits are straightforward bugfixes (e.g. the missing `MaybeFlipIPv6toCJDNS` calls or the logging improvements), whereas others are fixing some non-ideal behavior in some cases, but have the risk of introducing new non-ideal behavior in other cases because the fix comes with downsides of its own.

For example, if it isn't possible to distinguish situations in which two connections involving the same IP are controlled by the same entity or different entities, any solution would be imperfect, either allowing duplicate connections or preventing legitimate behavior (that is possibly widely used in local setups). So it's unclear to me if the status quo on these is actually a bug or an intended best-effort solution / the result of a compromise.

Maybe some of the straightforward bugfixes could be split out into another PR?

dergoegge commented on src/net.cpp:
```diff
@@ -1823,6 +1823,13 @@ void CConnman::CreateNodeFromAcceptedSocket(std::unique_ptr<Sock>&& sock,
 
     const bool inbound_onion = std::find(m_onion_binds.begin(), m_onion_binds.end(), addr_bind) != m_onion_binds.end();
+
+    // Do not accept inbound connections from I2P peers already connected to us,
+    // as creating new tunnels is expensive.
+    if (addr.IsI2P() && AlreadyConnectedToAddress(addr, ConnectionType::INBOUND)) {
+        return;
+    }
```
Isn't the tunnel already created at this point? Afaik the tunnel creation is handled by the i2p gateway (e.g. i2pd) not on our side.

I'm also not sure what we are protecting against here. Assuming tunnel creation is sufficiently DoSy (which it shouldn't be), this isn't enough to prevent attacks as an attacker could just not connect from the same address.

mzumsande commented:
> Note also, that item 7 in the PR description involves a regression in v26.x.

I don't think that this can be called a regression:
* There is nothing new about the root problem (making automatic connections to addnode peers in rare cases) in v26.
* It can happen that such a peer could now be protected as the only one of its network, but to determine if it's a regression this needs to be compared with the status before.
* Before we did protection-by-network, we would only ever get into the situation with an extra-outbound peer if we had a stale tip (30 minutes). Unless we are eclipsed or something weird is going on, this happens naturally a few times per day on average.
* If we are in that situation, it is not very probably that we'd evict the outbound peer that we'd rather have as an addnode peer. First of all, we'd usually just evict the new extra-peer (unless it provides a new block, which should happen very rarely, especially when we have other high-bandwidth OB peers at the same time). But even in the rare cases, we often would not evict the specific peer to solve the addnode problem, that peer might be one of the 4 protected peers, or we might just pick another one.

I'd guess you could probably have waited for months to actually evict a mis-placed peer that way. So I'd say that the stale-tip eviction was never a functioning mechanism to resolve these kind addnode problems (and never meant to be), and as a result, #27213 was not a regression - in practice, it might arguably be even an improvement instead of a regression, because in those cases where the peer in question is _not_ the only one from its network, we'll get into the situation where we have an extra peer to evict more frequently than before.

mzumsande commented:
> @mzumsande The improved logging allows distinguishing when this is due to the new protection logic, and with occasional exceptions, it is.

That's not my point. My point (see above) is that in earlier versions, the extra-outbound eviction (which would only happen in the stale-tip case then) was not a working or reliable mechanism to evict these peers, so there is no regression. Or did I miss something in my above explanation / do you have logs that pre-26 the stale-tip outbound eviction could somehow resolve this problem reliably?

mzumsande commented:
> Empirically, the regression is that we're actively adding addnode peers as outbound ones, and keeping them there. 

Oh, I thought you were talking about a race at startup - since the extra outbound should only kick in after 5 minutes, I think that should be enough to make connections to the addnode peers.

I think what might be happening later is that an addnoded peer disconnects us, and if it was the only one from its network then there could be a race between picking another network-related outbound peer (which could be the peer that just disconnected us) and reconnecting to the peer that evicted us with addnode thread. Could that be what you are seeing?

mzumsande commented:
> Maybe it would be prudent to allow feelers to addnodes.

I think it's fine to skip feelers for addnode peers. When we would skip these addresses for automatic connection, I don't see what making a feeler connection to them would achieve, considering that the point of feelers is to have better options for future automatic connections.

Waseem650 commented:
1884778242

Id
